File: 974. Subarray Sums Divisible by K.java
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        Map<Integer, Integer> prefixMap = new HashMap<>();
        int sum = 0;
        int answer = 0;
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            sum %= k;
            if (sum < 0) sum += k;
            if (sum == 0) answer++;
            if (prefixMap.containsKey(sum)) {
                answer += prefixMap.get(sum);
            }
            prefixMap.put(sum, prefixMap.getOrDefault(sum, 0) + 1);
        }
        return answer;
    }
}

File: 974. Subarray Sums Divisible by K.py
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        freq = {}
        freq[0] = 1
        answer, current = 0, 0
        for i in nums:
            current += i
            while current < 0:
                current += k
            current %= k
            if current not in freq:
                freq[current] = 0
            answer += freq[current]
            freq[current] += 1
        return answer
